File: clients/android/app/src/main/java/eu/allodia/mailcal/AllodiaSubscriptionWrites.kt
// The three things somebody may do to the subscription Allodia bills directly: stop the recurring
// charge, start it again, and move between the two periods.
//
// ⚠️ **Only Allodia's own subscription, never a store's.** A store's is changed at that store and
// nowhere else, so the card offers `manageUrl` for those and none of this. What decides whether a
// button exists at all is `actions`, which the service computed with every biller in view; a client
// is in no position to work that out and does not try.
//
// The fourth write, starting a checkout, is not here: it is what `WebBillingProvider` already calls
// behind the ordinary buy buttons, because on the `foss` flavour Allodia's own checkout *is* the
// shop (`purchasing.md`, "The shop follows the channel").
package eu.allodia.mailcal

import androidx.compose.material3.AlertDialog
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.platform.LocalContext
import java.util.Locale
import uniffi.mailcal_bindings.AllodiaOwnStatus
import uniffi.mailcal_bindings.AllodiaPlan
import uniffi.mailcal_bindings.AllodiaSubscription

// Which confirmation is on screen, if any.
//
// Both of these change what somebody is charged, and neither says so anywhere else: a cancellation
// is silent until the period runs out, and a switch is silent until a date that may be a month
// away. So both are asked before they are done.
private enum class AllodiaPendingWrite {
    CANCEL,
    SWITCH,
}

// The period a switch would move to, which is simply the other one.
//
// Null when there is nothing to switch: no subscription of Allodia's own, or a service that has
// said this account may not switch.
internal fun allodiaSwitchTarget(subscription: AllodiaSubscription): AllodiaPlan? {
    if (!subscription.actions.canSwitchInterval) return null
    return when (subscription.own?.interval) {
        AllodiaPlan.MONTHLY -> AllodiaPlan.YEARLY
        AllodiaPlan.YEARLY -> AllodiaPlan.MONTHLY
        null -> null
    }
}

// Whether to offer starting the subscription again, which is a different question from whether it
// has been cancelled: the service decides, because a cancelled subscription whose period has run
// out is a new checkout rather than a restart, and only it knows which.
internal fun allodiaOffersRestart(subscription: AllodiaSubscription): Boolean =
    subscription.actions.canResubscribe && subscription.own?.status == AllodiaOwnStatus.Cancelled

@Composable
internal fun AllodiaSubscriptionWrites(
    subscription: AllodiaSubscription,
    onCancel: () -> Unit,
    onResubscribe: () -> Unit,
    onSwitch: (AllodiaPlan) -> Unit,
) {
    val ctx = LocalContext.current
    // Screen-local on purpose, unlike the purchase state next door: neither of these leaves the
    // app, so a dialog dismissed by rotating the phone is a dialog nobody has answered yet.
    var pending by remember { mutableStateOf<AllodiaPendingWrite?>(null) }
    val switchTo = allodiaSwitchTarget(subscription)

    if (switchTo != null) {
        TextButton(onClick = { pending = AllodiaPendingWrite.SWITCH }) {
            Text(
                when (switchTo) {
                    AllodiaPlan.YEARLY -> L10n.settings_subscription_switch_yearly(ctx)
                    AllodiaPlan.MONTHLY -> L10n.settings_subscription_switch_monthly(ctx)
                }
            )
        }
    }
    if (allodiaOffersRestart(subscription)) {
        TextButton(onClick = onResubscribe) {
            Text(L10n.settings_subscription_resubscribe(ctx))
        }
    }
    if (subscription.actions.canCancel) {
        TextButton(onClick = { pending = AllodiaPendingWrite.CANCEL }) {
            Text(L10n.settings_subscription_cancel(ctx))
        }
    }

    when (pending) {
        AllodiaPendingWrite.CANCEL ->
            ConfirmWrite(
                title = L10n.settings_subscription_cancel_title(ctx),
                // ⚠️ The date is the whole reassurance: somebody cancelling wants to know they are
                // not losing what they have already paid for. A cancellation with no date reads as
                // "it stops now", which is the one thing it does not do.
                body = L10n.settings_subscription_cancel_body(ctx, allodiaWriteDate(subscription)),
                confirm = L10n.settings_subscription_cancel(ctx),
                // **Never "Cancel" for the way out.** On this dialog that word is the thing being
                // asked about, so the button that does nothing has to say what it keeps.
                dismiss = L10n.settings_subscription_cancel_keep(ctx),
                destructive = true,
                onConfirm = {
                    pending = null
                    onCancel()
                },
                onDismiss = { pending = null },
            )
        AllodiaPendingWrite.SWITCH ->
            ConfirmWrite(
                title = L10n.settings_subscription_switch_title(ctx),
                // ⚠️ **No price here, deliberately.** What this subscriber is charged is not
                // today's list price, because a price change never reaches somebody who already
                // subscribed, and quoting the list price would tell a long-standing subscriber a
                // number they will not be charged. The amount comes back from the switch itself
                // and is said afterwards.
                body = L10n.settings_subscription_switch_body(ctx, allodiaWriteDate(subscription)),
                confirm = L10n.action_update(ctx),
                dismiss = L10n.action_cancel(ctx),
                destructive = false,
                onConfirm = {
                    pending = null
                    switchTo?.let(onSwitch)
                },
                onDismiss = { pending = null },
            )
        null -> Unit
    }
}

@Composable
private fun ConfirmWrite(
    title: String,
    body: String,
    confirm: String,
    dismiss: String,
    destructive: Boolean,
    onConfirm: () -> Unit,
    onDismiss: () -> Unit,
) {
    AlertDialog(
        onDismissRequest = onDismiss,
        title = { Text(title) },
        text = { Text(body) },
        confirmButton = {
            TextButton(
                onClick = onConfirm,
                colors =
                    if (destructive) {
                        ButtonDefaults.textButtonColors(
                            contentColor = MaterialTheme.colorScheme.error
                        )
                    } else {
                        ButtonDefaults.textButtonColors()
                    },
            ) {
                Text(confirm)
            }
        },
        dismissButton = { TextButton(onClick = onDismiss) { Text(dismiss) } },
    )
}

// The day both confirmations talk about: what has already been paid for.
//
// The raw string when it cannot be read, which is the same fallback the card's own dates take: a
// sentence with the wire in it is still true, and one with no date at all is not.
@Composable
private fun allodiaWriteDate(subscription: AllodiaSubscription): String {
    val locale: Locale = allodiaReaderLocale()
    val raw = subscription.own?.currentPeriodEnd ?: subscription.currentPeriodEnd ?: return ""
    return allodiaDate(raw, locale) ?: raw
}

File: clients/apple/Packages/MailcalKit/Sources/MailcalUI/AllodiaSubscriptionWrites.swift
// What may be done to the subscription Allodia bills directly, and what each ending is called.
//
// ⚠️ **Only Allodia's own subscription, never a store's.** One bought through a store is cancelled,
// switched and refunded at that store, so the section offers its manage page for those and none of
// this. What decides whether a button exists at all is `actions`, which the service computed with
// every biller in view; a client is in no position to work that out and does not try.
//
// ⚠️ **Apple draws three of the four writes, and never the checkout.** Starting one means sending
// somebody to a payment page outside the app, which needs Apple's external-purchase-link
// entitlement, a declared target and Apple's own disclosure sheet. None of that is in place, so on
// this platform the store's own price stands alone and buying is StoreKit's. `purchasing.md` is
// where that rule lives.
//
// SwiftUI-free and pure on purpose, like the billing decisions next door: which sentence follows
// which ending is then a test rather than something only a live subscription can show. The twins
// are `AllodiaWriteResult.cs`, `AllodiaSubscriptionModel.kt` and `allodia_subscription_facts.rs`;
// keep the wording in step.

import Foundation
import MailcalBindings

/// How one write ended.
enum AllodiaWriteResult: Equatable {
    /// The recurring charge was stopped. Access runs to the date.
    case cancelled(endDate: String)
    /// Running again on the authorisation already held: nothing to pay, nothing to re-enter.
    case reactivated
    /// The next charge moved to the other period. Nothing moved today.
    case switched(AllodiaIntervalChange)
    /// Nothing to say. The service answered a restart with a payment page rather than a
    /// reactivation, and this platform may not open one, so the re-read is what reports where the
    /// subscription actually stands.
    case silent
    /// It did not go through, and the reason decides which sentence is owed.
    case failed(AllodiaWriteFailure)
}

/// Why a write did not happen, as the sentence to put in front of the person.
///
/// ⚠️ **A code, never a message from anywhere else.** The core reports a refusal as an
/// `AllodiaSubscriptionRefusal` and `purchasing.md` asks a client to switch on it, because "you
/// have already cancelled" and "that cannot change while a charge is being retried" are different
/// things to say and only the code tells them apart. Rendering the error instead puts a generated
/// description on screen: `errorDescription` is `String(reflecting:)`, so a refusal reads
/// `Refused(reason: MailcalBindings.AllodiaSubscriptionRefusal.notSwitchable)`.
enum AllodiaWriteFailure: Equatable {
    /// Nothing more can be said: an outage, a refused token, a reason this build has never heard
    /// of. The sentence says only that nothing changed.
    case unexplained
    case alreadyCancelled
    case alreadyActive
    case alreadyOnInterval
    /// A charge is mid-retry, so the period cannot move underneath it.
    case notSwitchable
    case notFound
}

/// The sentence an error earns.
///
/// Everything that is not a refusal is `unexplained`, a service that could not be reached
/// included: nothing was learned, so there is nothing to explain, and the one thing worth saying
/// is that nothing changed.
func allodiaWriteFailure(_ error: Error) -> AllodiaWriteFailure {
    guard
        let purchase = error as? AllodiaPurchaseError,
        case let .Refused(reason) = purchase
    else { return .unexplained }
    switch reason {
    case .alreadyCancelled: return .alreadyCancelled
    case .alreadyActive: return .alreadyActive
    case .alreadyOnInterval: return .alreadyOnInterval
    case .notSwitchable: return .notSwitchable
    case .notFound: return .notFound
    // Billing not configured on this deployment, and a reason a later service invents, both leave
    // a client with nothing specific it could truthfully say.
    case .unavailable, .other: return .unexplained
    }
}

/// The period a switch would move to, which is simply the other one, or `nil` when there is
/// nothing to switch: no subscription of Allodia's own, or a service that has said this account
/// may not switch.
///
/// `actions` decides, and nothing else. Somebody the App Store is charging has a period too, and
/// switching that one through this API is exactly what the service refuses.
func allodiaSwitchTarget(_ subscription: AllodiaSubscription) -> AllodiaPlan? {
    guard subscription.actions.canSwitchInterval, let own = subscription.own else { return nil }
    switch own.interval {
    case .monthly: return .yearly
    case .yearly: return .monthly
    }
}

/// Whether to offer starting the subscription again, which is a different question from whether it
/// has been cancelled: a cancelled subscription whose period has run out is a new checkout rather
/// than a restart, and only the service knows which.
func allodiaOffersRestart(_ subscription: AllodiaSubscription) -> Bool {
    subscription.actions.canResubscribe && subscription.own?.status == .cancelled
}

/// The day both confirmations talk about: what has already been paid for, as the service sent it,
/// or the empty string when it sent none.
///
/// ⚠️ The date is the whole reassurance. Somebody cancelling wants to know they are not losing what
/// they have already paid for, and a cancellation with no date reads as "it stops now", which is
/// the one thing it does not do.
func allodiaWriteDate(_ subscription: AllodiaSubscription) -> String {
    subscription.own?.currentPeriodEnd ?? subscription.currentPeriodEnd ?? ""
}

/// Minor units and a currency as something a reader recognises.
///
/// ⚠️ A switch answers an amount with no currency beside it, so it is priced from the
/// subscription's own `prices.currency`, which is today's list currency rather than the one this
/// subscriber was charged in. The two differ only for somebody whose billing currency has since
/// changed, which the service does not currently do.
@MainActor func allodiaMinorUnits(_ minorUnits: Int64, currency: String?) -> String {
    let amount = Decimal(minorUnits) / 100
    guard let currency, !currency.isEmpty else { return "\(amount)" }
    allodiaAmountFormatter.currencyCode = currency
    return allodiaAmountFormatter.string(from: amount as NSDecimalNumber) ?? "\(amount)"
}

/// Built once: `NumberFormatter` loads ICU locale data on construction, which the date helpers next
/// door carry the same warning about.
@MainActor private let allodiaAmountFormatter: NumberFormatter = {
    let output = NumberFormatter()
    output.numberStyle = .currency
    return output
}()
